WordPress/Classes/ViewRelated/Stats/JetpackLoginViewController.swift

```diff
@@ -0,0 +1,546 @@
+        NotificationCenter.default.removeObserver(self, name: .UITextFieldTextDidChange, object: self.passwordTextField)
+        NotificationCenter.default.removeObserver(self, name: .UITextFieldTextDidChange, object: self.verificationCodeTextField)
+    }
+
+    func textFieldChanged(_ notification: Foundation.Notification) {
+        updateSignInButton()
+    }
+
+    // MARK: - Keyboard
+
+    func registerForKeyboardNotifications() {
+        NotificationCenter.default.addObserver(self, selector: #selector(keyboardWillShow(_ :)),
+                                               name: .UIKeyboardWillShow, object: nil)
+        NotificationCenter.default.addObserver(self, selector: #selector(keyboardWillBeHidden(_ :)),
+                                               name: .UIKeyboardWillHide, object: nil)
+    }
+
+    func deregisterFromKeyboardNotifications() {
+        NotificationCenter.default.removeObserver(self, name: .UIKeyboardWillShow, object: nil)
+        NotificationCenter.default.removeObserver(self, name: .UIKeyboardWillHide, object: nil)
+    }
+
+    func keyboardWillShow(_ notification: Foundation.Notification) {
+        self.scrollView.isScrollEnabled = true
+        var info = notification.userInfo!
+        let keyboardSize = (info[UIKeyboardFrameBeginUserInfoKey] as? NSValue)?.cgRectValue.size
+        let contentInsets = UIEdgeInsets(top: 0, left: 0, bottom: keyboardSize!.height, right: 0)
+        self.scrollView.contentInset = contentInsets
+        self.scrollView.scrollIndicatorInsets = contentInsets
+
+        guard let activeField = self.activeField else {
+            return
+        }
+        var aRect: CGRect = self.view.frame
+        aRect.size.height -= keyboardSize!.height
+        if !aRect.contains(activeField.frame.origin) {
+            self.scrollView.scrollRectToVisible(activeField.frame, animated: true)
+        }
+    }
+
+    func keyboardWillBeHidden(_ notification: Foundation.Notification) {
+        self.scrollView.contentInset = UIEdgeInsets.zero
+        self.scrollView.scrollIndicatorInsets = UIEdgeInsets.zero
+        self.scrollView.isScrollEnabled = false
+    }
+
+    // MARK: - UI Helpers
+
+    fileprivate func reloadInterface() {
+        updateMessage()
+        updateControls()
+    }
+
+    fileprivate func updateMessage() {
+        guard let jetPack = self.blog.jetpack else {
+            return
+        }
+
+        var message: String
+
+        if jetPack.isInstalled() {
+            if jetPack.isUpdatedToRequiredVersion() {
+                message = NSLocalizedString("Looks like you have Jetpack set up on your site.\nCongrats!\n" +
+                                            "Sign in with your WordPress.com credentials below to enable " +
+                                            "Stats and Notifications.", comment: "")
+            } else {
+                message = NSLocalizedString("Jetpack \(JetpackVersionMinimumRequired) or later is required " +
```

Comment:
This may not pass through localization properly - I think we have to pass these in as parameters still in order for localization to work right.